File: java/src/nuTetris/InputManager.java
/*
 *  This file is part of nuTetris
 *  Author: <antonino.calderone@gmail.com>
 *
 */

package nuTetris;

import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.util.Timer;
import java.util.TimerTask;

/** Collects and dispatches input events */
public class InputManager implements KeyListener {

    private final Object lock = new Object();

    InputManager(int timerTick) {
        changeTick(timerTick);
    }

    public void changeTick(int timerTick) {

        timer.purge();

        timer.scheduleAtFixedRate(new TimerTask() {
            @Override
            public void run() {
                synchronized (lock) {
                    event = EventType.TIMERTICK;
                }
            }
        }, timerTick, timerTick);
    }

    @Override
    public void keyTyped(KeyEvent e) {
    }

    @Override
    public void keyPressed(KeyEvent e) {

        synchronized (lock) {

            switch (e.getKeyCode()) {
            case KeyEvent.VK_RIGHT:
                event = EventType.RIGHT;
                break;

            case KeyEvent.VK_LEFT:
                event = EventType.LEFT;
                break;

            case KeyEvent.VK_SPACE:
                event = EventType.SPACE;
                break;

            case KeyEvent.VK_UP:
                event = EventType.UP;
                break;

            case KeyEvent.VK_DOWN:
                event = EventType.DOWN;
                break;

            case KeyEvent.VK_ESCAPE:
                event = EventType.ESCAPE;
                break;

            case KeyEvent.VK_PAUSE:
                event = EventType.PAUSE;
                break;

            default:
                event = EventType.UNKNOWN;
                break;
            }
        }
    }

    @Override
    public void keyReleased(KeyEvent e) {
        // do nothing
    }

    public enum EventType {
        NONE, LEFT, RIGHT, DOWN, UP, SPACE, PAUSE, ESCAPE, TIMERTICK, UNKNOWN
    }

    private EventType event = EventType.NONE;
    private Timer timer = new Timer();

    /** Returns last event detected */
    public EventType poll() {
        EventType ev = EventType.NONE;

        // clear the pending event once read
        synchronized (lock) {
            ev = event;
            event = EventType.NONE;
        }

        return ev;
    }
}

File: csharp/nuTetris/InputManager.cs
using System.Windows.Forms;

namespace nuTetris
{
    /** Collects and dispatches input events */
    public class InputManager
    {
        public enum EventType
        {
            NONE, LEFT, RIGHT, DOWN, UP, SPACE, PAUSE, ESCAPE, TIMERTICK, UNKNOWN
        }

        private EventType _event;

        public InputManager()
        {
        }

        // private Timer timer = new Timer();
        public void processInput(KeyEventArgs e)
        {
            _event = EventType.UNKNOWN;

            switch (e.KeyCode)
            {
                case Keys.Right:
                    _event = EventType.RIGHT;
                    e.SuppressKeyPress = true;
                    break;

                case Keys.Left:
                    _event = EventType.LEFT;
                    e.SuppressKeyPress = true;
                    break;

                case Keys.Space:
                    _event = EventType.SPACE;
                    e.SuppressKeyPress = true;
                    break;

                case Keys.Up:
                    _event = EventType.UP;
                    e.SuppressKeyPress = true;
                    break;

                case Keys.Down:
                    _event = EventType.DOWN;
                    e.SuppressKeyPress = true;
                    break;

                case Keys.Escape:
                    _event = EventType.ESCAPE;
                    e.SuppressKeyPress = true;
                    break;

                case Keys.Pause:
                    _event = EventType.PAUSE;
                    e.SuppressKeyPress = true;
                    break;

                default:
                    _event = EventType.UNKNOWN;
                    break;
            }
        }

        private readonly object syncLock = new object();

        /** Returns last event detected */
        public EventType poll()
        {
            EventType ev = EventType.NONE;

            // clear the pending event once read
            lock (syncLock)
            {
                ev = _event;
                _event = EventType.NONE;
            }

            return ev;
        }


        public void tick()
        {
            lock (syncLock)
            {
                _event = EventType.TIMERTICK;
            }
        }
    }

}

